src/frontend/src/optimizer/rule/intersect_to_semi_join_rule.rs

```diff
@@ -0,0 +1,74 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_common::types::DataType::Boolean;
+use risingwave_common::util::iter_util::ZipEqDebug;
+use risingwave_pb::plan_common::JoinType;
+
+use super::{BoxedRule, Rule};
+use crate::expr::{ExprImpl, ExprType, FunctionCall, InputRef};
+use crate::optimizer::plan_node::{LogicalIntersect, LogicalJoin, PlanTreeNode};
+use crate::optimizer::PlanRef;
+
+pub struct IntersectToSemiJoinRule {}
+impl Rule for IntersectToSemiJoinRule {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let logical_intersect: &LogicalIntersect = plan.as_logical_intersect()?;
+        let all = logical_intersect.all();
+        if all {
+            return None;
```

Comment:
I think current implementation is `intersect all` rather `intersect`, since the latter will remove all duplicated rows.

Reply:
Oh, I was attempting to implement only the non-duplicate `intersect`, but it seems I overlooked the distinct part. `intersect all` is not an easy job so we can leave it later.

Reply:
I think we just need to remove the agg to implement `intersect all`?

Reply:
> I think we just need to remove the agg to implement `intersect all`?

More complex. You can take a look at the rule in flink.

https://github.com/apache/flink/blob/3a4c11371e6f2aacd641d86c1d5b4fd86435f802/flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/RewriteIntersectAllRule.scala

Reply:
I see, we need to consider which side contains more duplicated rows